Mathlib/NumberTheory/EulerProduct/Basic.lean

```diff
@@ -0,0 +1,170 @@
+  (summable_and_hasSum_smoothNumbers_prod_primesBelow_tsum hf₁ hmul
+    (fun hp ↦ hsum.comp_injective <| Nat.pow_right_injective hp.one_lt) _).2.tsum_eq.symm
+
+/-- A version of `EulerProduct.summable_and_hasSum_smoothNumbers_prod_primesBelow_geometric`
+in terms of the value of the series. -/
+lemma prod_primesBelow_geometric_eq_tsum_smoothNumbers {f : ℕ →* F} (hsum : Summable f) (N : ℕ) :
+    ∏ p in N.primesBelow, (1 - f p)⁻¹ = ∑' m : N.smoothNumbers, f m := by
+  refine (summable_and_hasSum_smoothNumbers_prod_primesBelow_geometric ?_ N).2.tsum_eq.symm
+  exact fun {_} hp ↦ hsum.norm_lt_one hp.one_lt
+
+/-- The following statement says that summing over `N`-smooth numbers
+for large enough `N` gets us arbitrarily close to the sum over all natural numbers
+(assuming `f` is norm-summable and `f 0 = 0`; the latter since `0` is not smooth). -/
+lemma norm_tsum_smoothNumbers_sub_tsum_lt (hsum : Summable f) (hf₀ : f 0 = 0)
+    {ε : ℝ} (εpos : 0 < ε) :
+    ∃ N₀ : ℕ, ∀ N ≥ N₀, ‖(∑' m : ℕ, f m) - (∑' m : N.smoothNumbers, f m)‖ < ε := by
+  obtain ⟨N₀, hN₀⟩ :=
+    summable_iff_nat_tsum_vanishing.mp hsum (Metric.ball 0 ε) <| Metric.ball_mem_nhds 0 εpos
+  simp_rw [mem_ball_zero_iff] at hN₀
+  refine ⟨N₀, fun N hN₁ ↦ ?_⟩
+  convert hN₀ _ <| N.smoothNumbers_compl.trans fun _ ↦ hN₁.le.trans
+  simp_rw [← tsum_subtype_add_tsum_subtype_compl hsum N.smoothNumbers,
+    add_sub_cancel', tsum_eq_tsum_diff_singleton (N.smoothNumbers)ᶜ hf₀]
+
+open Filter Nat in
+/-- The *Euler Product* for multiplicative (on coprime arguments) functions.
+If `f : ℕ → F`, where `F` is a complete normed field, `f 0 = 0`,
+`f 1 = 1`, `f` is multiplicative on coprime arguments,
+and `‖f ·‖` is summable, then `∏' p : {p : ℕ | p.Prime}, ∑' e, f (p ^ e) = ∑' n, f n`.
+Since there are no infinite products yet in Mathlib, we state it in the form of
+convergence of finite partial products. -/
+-- TODO: Change to use `∏'` once infinite products are in Mathlib
+theorem euler_product (hf₀ : f 0 = 0) :
+    Tendsto (fun n : ℕ ↦ ∏ p in primesBelow n, ∑' e, f (p ^ e)) atTop (𝓝 (∑' n, f n)) := by
+  rw [Metric.tendsto_nhds]
+  intro ε εpos
+  simp only [Finset.mem_range, eventually_atTop, ge_iff_le]
+  obtain ⟨N₀, hN₀⟩ := norm_tsum_smoothNumbers_sub_tsum_lt hsum.of_norm hf₀ εpos
+  use N₀
+  convert hN₀ using 3 with m
+  rw [dist_eq_norm, norm_sub_rev]
+  congr 2
+  exact prod_primesBelow_tsum_eq_tsum_smoothNumbers hf₁ hmul hsum m
+
+open Filter Nat in
+/-- The *Euler Product* for completely multiplicative functions.
+If `f : ℕ →*₀ F`, where `F` is a complete normed field
+and `‖f ·‖` is summable, then `∏' p : {p : ℕ | p.Prime}, (1 - f p)⁻¹ = ∑' n, f n`.
+Since there are no infinite products yet in Mathlib, we state it in the form of
+convergence of finite partial products. -/
+-- TODO: Change to use `∏'` once infinite products are in Mathlib
+theorem euler_product_multiplicative {f : ℕ →*₀ F} (hsum : Summable fun x => ‖f x‖) :
```

Comment:
I don't like the name here, since (to number theorists) "multiplicative" usually means weakly, not completely, multiplicative.

Reply:
Which names for the two results would you suggest instead?

Reply:
Maybe one should become `eulerProduct_completely_multiplicative`?

Reply:
OK.